Hw08

justmasyan opened the pull request.

justmasyan committed:
[HW-08] Скопировал прошлое ДЗ

justmasyan committed:
[HW-08] Выполнение ДЗ 8

justmasyan committed:
[HW-08] Почистил Application

stvort commented on hw08/src/main/java/ru/otus/hw/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    @Transactional
+    public BookDto insert(String title, long authorId, Set<Long> genresIds) {
+        return save(System.currentTimeMillis(), title, authorId, genresIds);
```
Ну и `System.currentTimeMillis()` грозит коллизиями

stvort commented on hw08/src/main/java/ru/otus/hw/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+    private final BookConverter bookConverter;
+
+    @Override
+    @Transactional(readOnly = true)
```
Транзакции точно сработают?

stvort commented on hw08/src/main/java/ru/otus/hw/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    @Transactional
+    public void deleteById(long id) {
+        bookRepository.deleteById(id);
```
А комментарии?

justmasyan committed:
[HW-08] Исправления ДЗ 8

justmasyan committed:
[HW-08] Добавил тесты сервисов

justmasyan committed:
Merge branch 'hw08' of https://github.com/justmasyan/2025-01-otus-winter-Hrenov into hw08

stvort commented on hw08/src/main/resources/application.yml:
```diff
@@ -0,0 +1,20 @@
+spring:
```
Приложение не запустилось пока не прописал отключение команды шелл "version", версию flapdoodle базы и 

```
mongock:
  mongo-db:
    write-concern:
      journal: false
    read-concern: local
```

stvort commented on hw08/src/main/java/ru/otus/hw/repositories/CommentaryRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+public interface CommentaryRepository extends MongoRepository<Commentary, String> {
+
+    @Query("{ 'book.$id': :#{#bookId} }")
```
Если это убрать, то все продолжит работать

stvort commented on hw08/src/main/java/ru/otus/hw/repositories/CommentaryRepository.java:
```diff
@@ -0,0 +1,18 @@
+    @Query("{ 'book.$id': :#{#bookId} }")
+    List<Commentary> findAllByBookId(@Param("bookId") String bookId);
+
+    @DeleteQuery("{ 'book.$id': :#{#bookId} }")
```
И т.д.

stvort commented on hw08/src/test/java/ru/otus/hw/services/BookServiceImplTest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Test
+    @DirtiesContext
```
Каждый раз пересоздавать контекст, избыточно. Можно только после некоторых изменений или перед некоторыми чтениями

justmasyan committed:
[HW-08] Установил использование embedded mongodb только для тестов
 + создал общий контекст для тестов, чтобы он создался один раз и пересоздавался после модифицирующих операций

stvort commented on hw08/src/main/java/ru/otus/hw/Application.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.data.mongodb.repository.config.EnableMongoRepositories;
+
+@EnableMongock
+@EnableMongoRepositories
```
`@EnableMongoRepositories` не нужен. Он включается стартером

stvort commented on hw08/src/test/java/ru/otus/hw/services/BookServiceImplTest.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Test
+    @DirtiesContext
```
> Каждый раз пересоздавать контекст, избыточно. Можно только после некоторых изменений или перед некоторыми чтениями

Тут можно перед `findAll` и после `update`, но вообще можно попробовать темплейтом просто базу чистить
